lightning/src/ln/outbound_payment.rs

```diff
@@ -571,6 +605,84 @@ impl OutboundPayments {
 			best_block_height, logger, pending_events, &send_payment_along_path)
 	}
 
+	pub(super) fn send_payment_for_bolt12_invoice<R: Deref, ES: Deref, NS: Deref, IH, SP, L: Deref>(
+		&self, invoice: &Bolt12Invoice, payment_id: PaymentId, router: &R,
+		first_hops: Vec<ChannelDetails>, inflight_htlcs: IH, entropy_source: &ES, node_signer: &NS,
+		best_block_height: u32, logger: &L,
+		pending_events: &Mutex<VecDeque<(events::Event, Option<EventCompletionAction>)>>,
+		send_payment_along_path: SP,
+	) -> Result<(), RetryableSendFailure>
+	where
+		R::Target: Router,
+		ES::Target: EntropySource,
+		NS::Target: NodeSigner,
+		L::Target: Logger,
+		IH: Fn() -> InFlightHtlcs,
+		SP: Fn(SendAlongPathArgs) -> Result<(), APIError>,
+	{
+		let payment_hash = invoice.payment_hash();
+		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+			hash_map::Entry::Occupied(entry) if entry.get().is_awaiting_invoice() => {
+				*entry.into_mut() = PendingOutboundPayment::InvoiceReceived { payment_hash };
+			},
+			hash_map::Entry::Occupied(_) => return Err(RetryableSendFailure::DuplicatePayment),
+			hash_map::Entry::Vacant(_) => return Err(RetryableSendFailure::UnexpectedPayment),
+		};
+
+		#[cfg(feature = "std")] {
+			if invoice.is_expired() {
+				log_error!(logger, "Invoice expired, abandoning payment {}", &payment_id);
+				self.abandon_payment(payment_id, PaymentFailureReason::PaymentExpired, pending_events);
+				return Err(RetryableSendFailure::PaymentExpired);
+			}
+		}
+
+		let recipient_onion = RecipientOnionFields {
```

Comment:
Ok, latest push reuses `retry_payment_internal` which now can return an error. If I drop the return value, then the two cases where we won't send an `InvoiceError` are for an expired invoice or route not found. Happy to drop it if you prefer.

I haven't updated the logging yet, but will do so after making a decision on the above to avoid conflicts.